[3주차] 황지영 / [feat] 게시글 CRUD 관련 기능 구현

hwangjiyoung02 opened the pull request:
## 1. 무슨 이유로 코드를 변경했나요?

- 썸네일의 경우 내용의 첫 이미지를 추출하여 사용했습니다.
- 조회의 경우 tokenfilter에서 제외를 하여 검증이 필요 없도록 하였습니다.
- 원래 이미지 도메인이 있었는데 삭제하고 순서를 고려하고자 PostContent에서 글과 함께 관리하도록 했습니다.

## 2. 어떤 위험이나 장애를 발견했나요?

- PostListResponse,PostResponse에서 반환하는 필드가 달라 dto를 구분했는데 똑같은 메서드를 중복으로 사용하여 static으로 만들었으나 내부에서 사용되는 메서드가 비static이여서 오류가 발생하였습니다. 여러 클래스에서 메서드가 중복으로 사용될 때 어떻게 처리하는지 궁금합니다.
- 저번에 생성자가 많으니까 이름을 다르게? 하시라고 한것 같은데 `public static Post of(Long userId, PostRequest postRequest) {    return new Post(userId, postRequest.title());}` 이런식으로 만드는게 맞는지 궁금합니다.

## 3. 관련 스크린샷을 첨부해주세요.

### 게시글작성

- 로그인 안했을 경우
    
![스크린샷 2025-04-06 155508](https://github.com/user-attachments/assets/63b82161-a886-46f6-9c89-737163aec466)

 - 로그인 했을 경우(=본인일 경우)
    
![스크린샷 2025-04-06 160550](https://github.com/user-attachments/assets/131fd690-fb89-4157-8cac-8210d7c9fae2)
![스크린샷 2025-04-06 160655](https://github.com/user-attachments/assets/008cfc8f-a74b-4c72-a2a2-74fdca2de7f4)
![스크린샷 2025-04-06 160714](https://github.com/user-attachments/assets/b70e3904-76fd-41e2-ad56-20c779ee6807)

    
        

### 게시글 조회

- 게시글 상세 조회
![스크린샷 2025-04-06 161553](https://github.com/user-attachments/assets/6c802020-ef39-4ccc-9a67-06794e8923ed)

    
 
- 게시글 전체 조회
    
![스크린샷 2025-04-07 012340](https://github.com/user-attachments/assets/7bdaf6a1-ca98-4946-b597-7d314c48b329)


### 게시글 수정

- 다른 유저 수정 시 접근 권한 오류 발생
![스크린샷 2025-04-07 023138](https://github.com/user-attachments/assets/f061f989-28cb-41a8-a3e4-4195157aef8a)

- 결과
![스크린샷 2025-04-06 163912](https://github.com/user-attachments/assets/12ace148-1ec6-4a05-a7e1-a8933e0a32af)

![스크린샷 2025-04-07 023332](https://github.com/user-attachments/assets/f2fb2190-d286-477a-a8b2-cfb9f288d408)

    
    

### 게시글 삭제

- 다른 유저 삭제 시 접근 권한 오류 발생
![스크린샷 2025-04-07 023354](https://github.com/user-attachments/assets/77a12da0-bf30-4181-9c18-6bc549d0357e)


- 삭제
    
    
![스크린샷 2025-04-06 164105](https://github.com/user-attachments/assets/cca19fb7-2cad-4561-b5d5-f260e8af8655)


## 4. 완료 사항
-  게시물 생성, 조회, 목록 조회, 수정, 삭제

## 5. 관련된 이슈

closed #38

hwangjiyoung02 committed:
🔨chore : 환경 설정 수정
- 스키마 및 도메인 수정
- 게시글 관련 도메인 생성
- BaseDomain 생성자 구현
- TokenFilter 조회 허용
- ErrorCode 추가

hwangjiyoung02 committed:
⚒️ refactoring : 인증 인가 코드 수정

hwangjiyoung02 committed:
✨feat : PostRepository 구현

hwangjiyoung02 committed:
✨feat : PostContent로 분리
- PostContentDto 생성
- PostContentRepository 생성

hwangjiyoung02 committed:
✨feat : PostController 및 dto 생성

hwangjiyoung02 committed:
✨feat : PostService 구현
- 생성
- 상세, 전체 조회
- 수정, 삭제

rootTiket commented on src/main/java/com/blog/domain/post/controller/dto/response/PostListResponse.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.stream.Collectors;
+
+public record PostListResponse(
+        Long postId,
```
Wrapper type을 사용하신 이유가 있으실까요?! id는 Null이 아님을 보장할 수 있게 원시타입도 고려해보시면 좋을것 같아요:):)

rootTiket commented on src/main/java/com/blog/domain/post/controller/dto/response/PostListResponse.java:
```diff
@@ -0,0 +1,53 @@
+        LocalDate createdAt
+) {
+
+    public static PostListResponse from(Post post, List<PostContent> postContents) {
+        List<PostContentDto> postContentDtos = toPostContentDtos(postContents);
+        String thumbnailUrl = extractThumbnailUrl(postContentDtos);
+        return new PostListResponse(
+                post.getId(),
+                post.getTitle(),
+                postContentDtos,
+                post.getCommentCount(),
+                thumbnailUrl,
+                post.getCreatedAt()
+        );
+    }
+
+    // postContents에서 postContentDto로 변환
+    private static List<PostContentDto> toPostContentDtos(List<PostContent> postContents) {
+        return postContents.stream()
+                .map(content -> new PostContentDto(
+                        ContentType.valueOf(content.getType().name()),
+                        content.getContent()
+                ))
+                .collect(Collectors.toList());
+    }
+
```
아마 PR에서 말씀해주신 부분이 이 부분인것 같네요! 
제 경우는 PostContentDto에서 static 메서드로 구현하고, from 메서드에서 `List<PostContentDto> contentDtos` 를 받는 식으로 구현할 것 같습니다 !!


```suggestion
    public static PostListResponse from(Post post, List<PostContentDto> postContentDtos) {
        String thumbnailUrl = extractThumbnailUrl(postContentDtos);
        return new PostListResponse(
                post.getId(),
                post.getTitle(),
                postContentDtos,
                post.getCommentCount(),
                thumbnailUrl,
                post.getCreatedAt()
        );
    }

```

rootTiket commented on src/main/java/com/blog/domain/post/controller/dto/response/PostResponse.java:
```diff
@@ -0,0 +1,63 @@
+
+) {
+
+    public static PostResponse from(Post post, List<PostContent> postContent, User user, boolean isOwner) {
+        List<PostContentDto> postContents = toPostContentDtos(postContent);
+        String thumbnailUrl = extractThumbnailUrl(postContents);
+        AuthorSummary authorSummary = toAuthorSummary(user);
+
+        return new PostResponse(
+                post.getId(),
+                post.getTitle(),
+                postContents,
+                post.getCommentCount(),
+                isOwner,
+                thumbnailUrl,
+                post.getCreatedAt(),
+                authorSummary
+        );
+    }
+
+    // postContents에서 postContentDto로 변환
+    private static List<PostContentDto> toPostContentDtos(List<PostContent> postContent) {
+        return postContent.stream()
+                .map(content -> new PostContentDto(
+                        ContentType.valueOf(content.getType().name()),
+                        content.getContent()
+                ))
+                .collect(Collectors.toList());
+    }
+
```
여기도 마찬가지로 ContentDto의 생성 책임을 해당 Dto로 넘기고, 서비스 로직에서 인자로 받아 PostResponse를 생성하는 책임만 남기면 중복 코드를 최소화 할 수 있을것 같아요 :):)

rootTiket commented on src/main/java/com/blog/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,149 @@
+    }
+
+    // 개별 조회
+    public PostResponse getPost(Long userId, Long postId) {
+        User user = validateUser(userId);
+        Post post = validatePost(postId);
+        List<PostContent> postContent = postContentRepository.findByPostIdOrderBySequence(postId);
+
+        boolean isOwner = post.getUserId().equals(userId);// 본인 글인지 판단
+        return PostResponse.from(post, postContent, user, isOwner);
+    }
+
+
+    // 전체 조회
+    public List<PostListResponse> getPostList(Long userId) {
+        List<Post> posts = postRepository.findAllByUserId(userId);
+
+        // 유저가 쓴 글만 가져오기
+        return posts.stream()
+                .map(post -> {
+                    List<PostContent> postContents = postContentRepository.findByPostIdOrderBySequence(post.getId());
+                    return PostListResponse.from(post, postContents);
+                })
+                .collect(Collectors.toList());
+    }
+
```
여기서 ContentDto를 생성해서 from에 넘겨주는 방식이에요 !! (제가 너무 어렵게 설명 했을까봐 ㅠ...)

hwangjiyoung02 replied:
아니요! 잘 이해했습니다:) 친절한 설명 감사드려요~

rootTiket reviewed (approved):
고생하셨습니다!!! 지난번에 record 추천해드렸는데 반영하셔서 뿌듯하네요 :) 질문 작성해주신 부분에 대한 답글도 달아두었느니 참고해주시면 감사할것 같아요 !!!

soyesenna commented on src/main/java/com/blog/domain/post/controller/PostController.java:
```diff
@@ -0,0 +1,79 @@
+    // 글 등록
+    @PostMapping
+    public ApiResponse<String> createPost(HttpServletRequest request, @Valid @RequestBody PostRequest postRequest) {
+        Long userId = tokenService.getUserIdFromCookie(request);
```
인가 로직이 컨트롤러 전반에 중복된 것 처럼 보여요!
AOP, Filter 등을 활용해서 추상화해보시는건 어떤가요?

soyesenna commented on src/main/java/com/blog/global/common/BaseDomain.java:
```diff
@@ -1,10 +1,29 @@
 
-    private LocalDateTime createdAt;
-    private LocalDateTime updatedAt;
+    protected LocalDate createdAt;
+    protected LocalDate updatedAt;
```
`createAt`, `updatedAt`을 `LocalDate`로 바꾸신 이유가 있으실까요??

soyesenna reviewed (approved):
구현 고생하셨어요!!
몇 가지 궁금한 점 적어두었습니다😊

hwangjiyoung02 committed:
🔥fix : cookie에서 id 추출 로직 - ArgumentResolver로 구현
- GetUserId 인터페이스 구현
- UserIdArgumentResolver로 cookie에서 userId추출하는 메서드 구현
- Webconfig에 등록

hwangjiyoung02 committed:
⚒️refactoring : 중복 로직 PostContentDto에 static 메서드 구현

hwangjiyoung02 committed:
⚒️refactoring : LocalDate->LocalDateTime으로 변경

hwangjiyoung02 commented:
AOP를 사용하려고 했는데 쿠키에서 userId 추출하는 로직은 메서드 실행하기 보다는 파라미터에 바인딩하는 방법인 ArgumentResolver방식이 나은 것 같아 적용해보았습니다!
